### PR TITLE
Typechecker

### LINKED ISSUE #76: (Code) Type / Scope - Build FTable
Define `SctTableBuilder.cs` in the `Sct.Compiler` namespace, which traverses the AST to create a valid FTable structure which can be continously populated with the currently in-scope functions, along with their types. 

The FTable may likely require another class `FTable.cs`.

This first requires figuring out a structure which can handle providing only in-scope functions to the type checker. This means it has to be dynamic in some way.
- Type checker may not know which class it has entered when entering, how do we handle this case?

Plan and discuss a strategy before implementing it.

### LINKED ISSUE #79: (Code) Type / Scope - Vtable
Create VTable stack class used for storing the current variable scope. Should include `Push` / `Pop` / `EnterScope` / `ExitScope`.

### LINKED ISSUE #86: (Code) Type - Build CTable
Like FTables introduced in SPO, we need a CTable to lookup agents and get a state.
Can be created by the same listener which creates FTables. See FTable ticket 
- Can be implemented as simply a dictionary indexed with class name, outputting set of state strings.

### LINKED ISSUE #107: (Code) Test - Type checker
Compare output from `SctReturnCheckVisitor`, with snapshots, and expect it to be either none, or some error(s) that the code contains.

Likewise with type checking.
________________________________________________________

Sidste ting der mangler:

Skal assignment INT = FLOAT være mulig?
Type mismatch forkert fejlbesked I compatibleType()?

- [x] NoReturn virker ikke
- [x] PredAgentDoesNotExist smider sgu lige exception :)
- [x] PredTypeMisMatch skal bruge getCompatibleType()
- [x] PredVariableDoesNotExist virker ikke
- [x] UnknownDecorator virker ikke
- [x] declaring void variable prints "Variable cannot be of type :Sct.Compiler.SctType"
- [x] visitReturn skal bruge getCompatible type right?
- [x] lav en none type I typetable så vi ikke skriver "return TypeTable.Void;"
- [ ] flyt add error ved conversion error ind I getcompatibletype?
- [ ] Lav noget så vi printer Int I stedet for System.Int32 for alle typer
- [x] fix alle test filer så de ikke indeholder expressions for sig selv fx. "7+x" skal være "x=7+x"

spørgsmål:
hvad er variableReused.sct i test?
er der brug for intToFloat.sct i test?

dotnet run: 
var A B og C findes ikke virker ikke

(test samme variablenavn I parameter)

### LINKED ISSUE #115: (code) Type - BooleanExpression

### LINKED ISSUE #117: (Code) Type - Check return type

### LINKED ISSUE #118: (Code) Type - Variable assignment

### LINKED ISSUE #119: (Code) Type - Check typecast

### LINKED ISSUE #121: (Code) Type - check if predicate in expressions
check if the following contain predicates

LiteralExpression
IDExpression
ParenthesisExpression
TypecastExpression
CallExpression
UnaryMinusExpression
LogicalNotExpression
boolean expressions

### LINKED ISSUE #122: (Code) Type - predicates

### LINKED ISSUE #145: (code) type - check for existing classes

### LINKED ISSUE #146: (Code) Type - Type unary minus expressions

### LINKED ISSUE #147: (Code) Type - type check if and if else

### LINKED ISSUE #148: (Code) Type - type check while

### PR DESCRIPTION
**Pull request by Thomast2001**
closes #86 closes #122 closes #146 closes #145 closes #118 closes #121 closes #147 closes #148 closes #115 closes #79 closes #119 closes #107 closes #117 closes #76

please review this very short pull request

### CODE REVIEW

File: SocietalConstructionTool/Compiler/Typechecker/SctTypeChecker.cs [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+            {
+                return TypeTable.Int;
+            }
+            else if (context.FLOAT() is not null)
+            {
+                return TypeTable.Float;
+            }
+            else
+            {
+                _errors.Add(new CompilerError("Literal type not recognized, must be int or float.", context.Start.Line, context.Start.Column));
```
Comment (TheMagzuz) [inline SocietalConstructionTool/Compiler/Typechecker/SctTypeChecker.cs:242]: Maybe also exception?

File: SocietalConstructionTool/Compiler/Typechecker/SctTypeChecker.cs [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+            var targetAgentFields = targetAgent.Fields;
+            var agentArgumentIds = context.args_agent().ID();
+
+            var agentArgs = new Dictionary<string, SctType>();
+
+            foreach (var id in agentArgumentIds)
+            {
+                if (!targetAgentFields.TryGetValue(id.GetText(), out _))
+                {
+                    _errors.Add(new CompilerError($"Variable {id.GetText()} does not exist in agent {agentName}", context.Start.Line, context.Start.Column));
```
Comment (TheMagzuz) [inline SocietalConstructionTool/Compiler/Typechecker/SctTypeChecker.cs:360]: Field instead of variable?

**Review by Bliztle — CHANGES_REQUESTED** [review]
There is a general inconsitency in the usage of TypeTable.None and TypeTable.Void, where a lot of voids should be none, for nodes returning no type.

File: SocietalConstructionTool/Compiler/Typechecker/VtableEntry.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    public class VtableEntry
+    {
+        public string Name { get; }
+        public SctType Type { get; }
+
+        public VtableEntry(string name, SctType type)
+        {
+            Name = name;
+            Type = type;
+        }
```
Comment (TheMagzuz) [inline SocietalConstructionTool/Compiler/Typechecker/VtableEntry.cs:12]: Primary constructor

**Comment by Tinggaard** [comment]
> There is a general inconsitency in the usage of TypeTable.None and TypeTable.Void, where a lot of voids should be none, for nodes returning no type.

This should be fixed now

**Review by Bliztle — APPROVED** [review]

**Review by Bliztle — APPROVED** [review]

**Pull request merged**